// element-angular/release/collapse/collapse.js
import { Component, EventEmitter, Input, Output } from '@angular/core';
export class ElCollapse {
    constructor() {
        this.modelValue = [];
        this.accordion = false;
        this.modelChange = new EventEmitter();
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set model(val) {
        this.modelValue = val;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    updateModel(value) {
        const /** @type {?} */ index = this.modelValue.findIndex(val => val === value);
        if (index < 0) {
            this.accordion && (this.model = []);
            this.modelValue.push(value);
            return this.modelChange.emit(this.modelValue);
        }
        if (this.accordion) {
            this.model = [];
        }
        else {
            this.modelValue.splice(index, 1);
        }
        this.modelChange.emit(this.modelValue);
    }
}
ElCollapse.decorators = [
    { type: Component, args: [{
                selector: 'el-collapse',
                template: `
    <div class="el-collapse" role="tablist" aria-multiselectable="true">
      <ng-content></ng-content>
    </div>
  `,
            },] },
];
/**
 * @nocollapse
 */
ElCollapse.ctorParameters = () => [];
ElCollapse.propDecorators = {
    'accordion': [{ type: Input },],
    'model': [{ type: Input },],
    'modelChange': [{ type: Output },],
};
function ElCollapse_tsickle_Closure_declarations() {
    /** @type {?} */
    ElCollapse.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ElCollapse.ctorParameters;
    /** @type {?} */
    ElCollapse.propDecorators;
    /** @type {?} */
    ElCollapse.prototype.modelValue;
    /** @type {?} */
    ElCollapse.prototype.accordion;
    /** @type {?} */
    ElCollapse.prototype.modelChange;
}
