// element-angular/release/cascader/cascader-props.js
import { EventEmitter, Input, Output } from '@angular/core';
export class ElCascaderPoprs {
    constructor() {
        this.elDisabled = false;
        this.placeholder = '请选择';
        this.clearable = false;
        this.allLevels = true;
        this.changeOnSelect = false;
        this.modelChange = new EventEmitter();
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set disabled(val) {
        console.warn('Element Angular: (disabled) is discarded, use [elDisabled] replace it.');
    }
}
ElCascaderPoprs.propDecorators = {
    'disabled': [{ type: Input },],
    'elDisabled': [{ type: Input },],
    'size': [{ type: Input },],
    'placeholder': [{ type: Input },],
    'options': [{ type: Input },],
    'clearable': [{ type: Input },],
    'allLevels': [{ type: Input, args: ['all-levels',] },],
    'changeOnSelect': [{ type: Input, args: ['change-on-select',] },],
    'model': [{ type: Input },],
    'modelChange': [{ type: Output },],
};
function ElCascaderPoprs_tsickle_Closure_declarations() {
    /** @type {?} */
    ElCascaderPoprs.propDecorators;
    /** @type {?} */
    ElCascaderPoprs.prototype.elDisabled;
    /** @type {?} */
    ElCascaderPoprs.prototype.size;
    /** @type {?} */
    ElCascaderPoprs.prototype.placeholder;
    /** @type {?} */
    ElCascaderPoprs.prototype.options;
    /** @type {?} */
    ElCascaderPoprs.prototype.clearable;
    /** @type {?} */
    ElCascaderPoprs.prototype.allLevels;
    /** @type {?} */
    ElCascaderPoprs.prototype.changeOnSelect;
    /** @type {?} */
    ElCascaderPoprs.prototype.model;
    /** @type {?} */
    ElCascaderPoprs.prototype.modelChange;
}
