// element-angular/release/tree/tree.js
import { Component } from '@angular/core';
import { ModelStandard } from './utils';
import { ElTreeProps } from './tree-props';
export class ElTree extends ElTreeProps {
    constructor() {
        super();
        this.userSafeHooks = () => ({
            findAllChecked: this.findAllChecked.bind(this),
            updateItemChecked: this.updateChecked.bind(this),
            updateItemExpanded: this.updateExpanded.bind(this),
            removeAllChecked: this.removeAllChecked.bind(this),
        });
    }
    /**
     * @return {?}
     */
    findAllChecked() {
        if (!this.showCheckbox)
            return [];
        return ModelStandard.FindAllChecked(this.identModel);
    }
    /**
     * @return {?}
     */
    removeAllChecked() {
        ModelStandard.LoopRemoveChecked(this.identModel);
    }
    /**
     * @param {?} id
     * @return {?}
     */
    updateExpanded(id) {
        this.identModel = ModelStandard.DeepUpdateExpanded(id, this.identModel, this.accordion);
    }
    /**
     * @param {?} id
     * @return {?}
     */
    updateChecked(id) {
        if (!this.showCheckbox)
            return;
        this.identModel = ModelStandard.DeepUpdateChecked(id, this.identModel);
    }
    /**
     * @param {?} next
     * @return {?}
     */
    emitter(next) {
        this.modelChange.emit(next);
    }
}
ElTree.decorators = [
    { type: Component, args: [{
                selector: 'el-tree',
                template: `
    <div class="el-tree" role="tree">
      <el-tree-item *ngFor="let item of identModel"
        [model]="item" [indent]="indent">
      </el-tree-item>
      <div class="el-tree__empty-block" *ngIf="!identModel && !identModel.length">
        <span class="el-tree__empty-text">{{ emptyText }}</span>
      </div>
    </div>
  `,
            },] },
];
/**
 * @nocollapse
 */
ElTree.ctorParameters = () => [];
function ElTree_tsickle_Closure_declarations() {
    /** @type {?} */
    ElTree.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ElTree.ctorParameters;
    /** @type {?} */
    ElTree.prototype.userSafeHooks;
}
